protege-distribution: refactor patching process and icon generation

nessdoor opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Following the discussion stimulated by the merge of https://github.com/NixOS/nixpkgs/pull/127396 in place of https://github.com/NixOS/nixpkgs/pull/97365, I included in this PR the following suggested improvements:
- more stable and self-documenting patching process
- automatic production of icon images from the upstream `.ico` file

Refer to https://github.com/NixOS/nixpkgs/pull/127396#issuecomment-864947179, https://github.com/NixOS/nixpkgs/pull/127396#issuecomment-865119400 and https://github.com/NixOS/nixpkgs/pull/127396#issuecomment-865365527 for the full rationale.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

nessdoor committed:
protege-distribution: rewrite patching mechanism

Part of the launcher script and configuration were patched via sed
commands; make the process clearer by using unified-style patches.

nessdoor committed:
protege-distribution: auto-generate all icon sizes for theme hicolor

Previous iteration simply copied the upstream 128x128 .ico into the
icon search path; use icoFileToHiColorTheme to produce the complete
tree, instead.

nessdoor commented:
@mucaho I saw that your `.desktop` file specification was more involved than mine. If you feel the need, please show me what meaningful changes I can make.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 128027` at 810f43cd run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>protege-distribution</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`lgpl3` is a deprecated license, please check if project uses `lgpl3Plus` or `lgpl3Only` and change `meta.license` accordingly.

Near pkgs/development/web/protege-distribution/default.nix:72:5:
```
   |
72 |     license = with licenses; [ asl20 bsd2 epl10 lgpl3 ];
   |     ^
```
</li>
  </ul>
</details>

mucaho reviewed (changes requested):
##### Reviewed points

- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package build on Nixos 20.03.2881.5607e74b1dd (Markhor)
- [x] executables tested on x64
- [x] all depending packages build - there are none

##### Possible improvements

Please extend desktop file specification

##### Comments

Thank you very much for cleaning this package up :+1:

nessdoor committed:
protege-distribution: extend desktop file for grouping

Specifying 'categories' allows Protege to be consistently grouped
inside application launcher submenus.

nessdoor commented on pkgs/development/web/protege-distribution/default.nix:
```diff
@@ -56,7 +58,7 @@ stdenv.mkDerivation rec {
       name = "Protege";
       desktopName = "Protege Desktop";
-      icon = "protege.ico";
+      icon = "protege";
```
I committed the StartupWMClass modification, but my graphical environment doesn't pick the additional config up (probably because of its non-compliance). Please, let me know if it looks correct on your end.

mucaho replied:
I think you need to logout of your graphical session and log back in again, that updates the menu entries for me at least.

I installed the new version of protege and it was placed into the appropriate sub-menu.

However, I can not verify the window grouping will work correctly, as I don't have window grouping in xfce and I still get `> WM_CLASS(STRING) = "sun-awt-X11-XFramePeer", "java-lang-Thread"` despite the `StartupWMClass` change.

According to [this SO answer](https://askubuntu.com/a/367851): _"So the field called StartupWMClass is a string that defines in which class the application is grouped in the Unity Launcher at startup."_
However, as seen [in this SO post](https://askubuntu.com/a/741940): _"StartupWMClass does not change the actual application window class, but only tells the desktop manager what window class should use to identify the application window(s)."_
On the other hand [this SO user claims](https://askubuntu.com/questions/653702/startupwmclass-will-not-change-wm-class-of-eclipse-ide#comment939567_653702): _"Same behavior here with my both eclipse installations (Java, C++). xprop WM_CLASS gives WM_CLASS(STRING) = "Eclipse", "Eclipse" in both cases [when using different StartupWMClasses]. But my running instances are not grouped."_

Alas, I'm quite puzzled now.

mucaho replied:
Actually, I found the option for grouping applications. After conducting multiple tests, it does not work as expected. `StartupWMClass` has no effect on the grouping, at least in xfce, as the WM_Class still says "java-lang-Thread" - that's also the name of the group created.

The only proper solution I have found is to fix this upstream, in protege itself, e.g. see https://stackoverflow.com/a/29218320.

Unless someone else has a better idea, I'd suggest to remove that `StartupWMClass` entry, since it doesn't achieve anything apparently.
Sorry for bothering you with this, it seems I was missing a piece copying the desktop configuration from IntelliJ.

nessdoor replied:
Alright, removed and re-committed, anything else?

mucaho reviewed (approved):
Thanks again, retested the application, LGTM :+1:

mucaho commented:
pinging @SuperSandro2000 , if you could please take a look and merge if it's ok
